zhusuan/distributions/special.py

```diff
@@ -0,0 +1,168 @@
+
+    def _prob(self, given):
+        raise ValueError("An empirical distribution has no probability measure.")
+
+
+class Implicit(Distribution):
+    """
+    The class of Implicit distribution.
+    See :class:`~zhusuan.distributions.base.Implicit` for details.
+    :param name: A string. The name of the `StochasticTensor`. Must be unique
+        in the `BayesianNet` context.
+    :param implicit: A N-D (N >= 1) `float` Tensor
+    :param value_shape: A list or tuple describing the `value_shape` of the distribution.
+        The entries of the list can either be int, Dimension or None.
+    :param group_ndims: A 0-D `int32` Tensor representing the number of
+        dimensions in `batch_shape` (counted from the end) that are grouped
+        into a single event, so that their probabilities are calculated
+        together. Default is 0, which means a single value is an event.
+        See :class:`~zhusuan.distributions.base.Distribution` for more detailed
+        explanation.
+    """
+
+    def __init__(self,
+                 implicit,
+                 value_shape=None,
+                 group_ndims=0,
+                 **kwargs):
+        self.implicit = implicit
+
+        if value_shape is None:
+            self.explicit_value_shape = []
+        elif not isinstance(value_shape, (list, tuple)):
+            self.explicit_value_shape = [value_shape]
+        else:
+            self.explicit_value_shape = list(value_shape)
+
+        super(Implicit, self).__init__(
+            dtype=implicit.dtype,
+            param_dtype=implicit.dtype,
+            is_continuous=implicit.dtype.is_floating,
+            group_ndims=group_ndims,
+            is_reparameterized=False,
+            **kwargs)
+
+    def _value_shape(self):
+        return tf.convert_to_tensor(self.explicit_value_shape, tf.int32)
+
+    def _get_value_shape(self):
+        return tf.TensorShape(self.explicit_value_shape)
+
+    def _batch_shape(self):
+        d = len(self.explicit_value_shape)
+        if d == 0:
+            return tf.shape(self.implicit)
+        else:
+            return tf.shape(self.implicit)[:-d]
+
+    def _get_batch_shape(self):
+        if self.implicit.get_shape():
+            d = len(self.explicit_value_shape)
+            if d == 0:
+                return self.implicit.get_shape()
+            else:
+                return self.implicit.get_shape()[:-d]
+        return tf.TensorShape(None)
+
+    def _sample(self, n_samples=None):
+        if n_samples is not None and n_samples != 1:
+            raise ValueError("ImplicitDistribution does not accept `n_samples` argument.")
+        return tf.expand_dims(self.implicit, 0)
+
+    def _log_prob(self, given):
+        return tf.log(self.prob(given))
+
+    def _prob(self, given):
+        given = tf.cast(given, self.param_dtype)
+        given, implicit = maybe_explicit_broadcast(given, self.implicit, 'given', 'implicit')
+        prob = tf.cast(tf.equal(given, implicit), tf.float32)
```

Comment:
The two casts are unnecessary. Since given will be checked to be compatible with self.dtype.